File: crates/yamind-core/src/boundary.rs
use crate::id::{BoundaryId, NodeId};
use crate::style::Color;
use serde::{Deserialize, Serialize};

/// A visual grouping frame around a set of nodes.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Boundary {
    pub id: BoundaryId,
    pub label: String,
    pub node_ids: Vec<NodeId>,
    pub fill_color: Color,
    pub stroke_color: Color,
    pub stroke_width: f32,
    pub padding: f32,
}

impl Boundary {
    pub fn new(node_ids: Vec<NodeId>) -> Self {
        Self {
            id: BoundaryId::new(),
            label: String::new(),
            node_ids,
            fill_color: Color::rgba(0.3, 0.5, 0.8, 0.1),
            stroke_color: Color::rgba(0.45, 0.65, 0.95, 0.7),
            stroke_width: 1.5,
            padding: 10.0,
        }
    }
}

File: yamindmap-electron/src/shared/types/boundary.ts
import type { Color } from './style'
import type { NodeId } from './node'

export type BoundaryId = string

export interface Boundary {
  id: BoundaryId
  label: string
  show_label: boolean
  node_ids: NodeId[]
  fill_color: Color
  stroke_color: Color
  stroke_width: number
  padding: number
}

export const BOUNDARY_DEFAULTS = {
  fill_color: { r: 0.3, g: 0.5, b: 0.8, a: 0.1 },
  stroke_color: { r: 0.45, g: 0.65, b: 0.95, a: 0.7 },
  stroke_width: 1.5,
  padding: 10.0
} as const

export function createBoundary(id: BoundaryId, nodeIds: NodeId[]): Boundary {
  return {
    id,
    label: '',
    show_label: true,
    node_ids: nodeIds,
    fill_color: { ...BOUNDARY_DEFAULTS.fill_color },
    stroke_color: { ...BOUNDARY_DEFAULTS.stroke_color },
    stroke_width: BOUNDARY_DEFAULTS.stroke_width,
    padding: BOUNDARY_DEFAULTS.padding
  }
}
